File: solutions/cpp/P10_SummationOfPrimes.cpp
#include <iostream>
#include "euler_util\primes.hpp"

unsigned long long summation_of_primes_to(unsigned long long max) {
    unsigned long long sum = 0;
    using primes::primes;
    for (auto p = primes.begin(); *p < max; ++p) {
      sum += *p;
    }
    return sum;
}

int main() {
    unsigned long long max = 2000000;
    std::cout << "Solution: " << summation_of_primes_to(max) << std::endl;
    return 0;
}

File: solutions/haskell/P10_SummationOfPrimes.hs
import System.Environment (getArgs)
--import EulerUtil.Primes (primesTo)

primesTo :: Int -> [Int]
primesTo m
    | m <  2 = []
    | m == 2 = [2]
    | otherwise = 2 : sieve [3,5..m]
        where sieve [] = []; sieve (x:xs) = x : sieve (filter (\n-> mod n x /= 0) xs)

main = do
    arguments <- getArgs
    let n = read $ head arguments
    let result = sum $ primesTo n
    putStrLn $ show result
